src/TestResults/match-by-context.ts

```diff
@@ -0,0 +1,362 @@
+    if (another.zeroBasedLine !== this.zeroBasedLine) {
+      return false;
+    }
+    this.data.push(...another.data);
+    return true;
+  }
+
+  /** return the only element in the list, exception otherwise */
+  only(): T {
+    if (this.data.length !== 1) {
+      throw new TypeError(`expect 1 element but got ${this.data.length} elements`);
+    }
+    return this.data[0];
+  }
+  /** return the first element, if no element, returns undefined */
+  first(): T | undefined {
+    if (this.data.length > 0) {
+      return this.data[0];
+    }
+  }
+}
+
+export class ContainerNode<T> implements BaseNodeType {
+  public childContainers: ContainerNode<T>[] = [];
+  public childData: DataNode<T>[] = [];
+  public zeroBasedLine: number;
+  public name: string;
+
+  constructor(name: string) {
+    this.name = name;
+    this.zeroBasedLine = -1;
+  }
+
+  public addContainerNode(container: ContainerNode<T>): void {
+    this.childContainers.push(container);
+  }
+
+  public addDataNode(dataNode: DataNode<T>): void {
+    this.childData.push(dataNode);
+  }
+
+  public findContainer(path: string[], createIfMissing = true): ContainerNode<T> | undefined {
+    if (path.length <= 0) {
+      return this;
+    }
+    const pn = path[0];
+    let container = this.childContainers.find((c) => c.name === pn);
+    if (!container && createIfMissing) {
+      container = new ContainerNode(pn);
+      this.addContainerNode(container);
+    }
+    return container?.findContainer(path.slice(1), createIfMissing);
+  }
+
+  /**
+   * deeply sort all child-data and child-containers by line position.
+   * @param grouping if true, will try to merge child-data with the same line
+   */
+  public sort(grouping = false): void {
+    const sortByLine = (n1: BaseNodeType, n2: BaseNodeType): number =>
+      n1.zeroBasedLine - n2.zeroBasedLine;
+    const groupData = (list: DataNode<T>[], data: DataNode<T>): DataNode<T>[] => {
+      if (list.length <= 0) {
+        return [data];
+      }
+      // if not able to merge with previous node, i.e . can not group, add it to the list
+      if (!list[list.length - 1].merge(data)) {
+        list.push(data);
+      }
+      return list;
+    };
+
+    this.childData.sort(sortByLine);
+    if (grouping) {
+      this.childData = this.childData.reduce(groupData, []);
+    }
+
+    // recursive to sort childContainers, then sort the list itself
+    this.childContainers.forEach((c) => c.sort(grouping));
+    this.childContainers.sort(sortByLine);
+
+    // if container doesn't have valid line info, use the first child's
+    if (this.zeroBasedLine < 0) {
+      const topLines = [this.childData, this.childContainers]
+        .map((l) => (l.length > 0 ? l[0].zeroBasedLine : undefined))
+        .filter((n) => n);
+      this.zeroBasedLine = Math.min(...topLines);
+    }
+  }
+}
+type NodeType<T> = ContainerNode<T> | DataNode<T>;
+
+const ROOT_NODE_NAME = '__root__';
+export const buildAssertionContainer = (
```

Comment:
you mean besides my personal preference of arrow-functions 😏 ? Unlike class function declaration that can take advantage of the prototype, top-level export function vs const do not make much difference, however, since jest has problem mocking exported functions, more reason to use const here...

Reply:
Ah thanks, that's also a good argument.